File: backend/scripts/upload_parallel.py
import os
import subprocess
import concurrent.futures
import sys
import time

# Configuration
DATA_DIR = os.path.join(os.path.dirname(__file__), '../../data/chunks')
BUCKET_NAME = 'austin-codes'
CONCURRENCY = 10

def get_files(directory):
    file_list = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            file_list.append(os.path.join(root, file))
    return file_list

def upload_file(file_path):
    # Calculate relative path for the key
    # We want the key to be relative to 'data/' folder, e.g. 'chunks/manifest.json'
    # The DATA_DIR is '.../data/chunks', so we go up one level to '.../data'
    base_data_dir = os.path.abspath(os.path.join(DATA_DIR, '..'))
    relative_key = os.path.relpath(file_path, base_data_dir).replace('\\', '/')
    
    cmd = ['raindrop', 'object', 'put', file_path, relative_key, '--bucket', BUCKET_NAME]
    
    # Run command from backend directory so it finds raindrop.manifest
    backend_dir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
    # shell=True is required on Windows to find the 'raindrop' command (which is likely a .cmd file)
    result = subprocess.run(cmd, capture_output=True, text=True, shell=True, cwd=backend_dir)
    
    if result.returncode != 0:
        return False, f"Failed to upload {relative_key}: {result.stderr}"
    return True, relative_key

def main():
    print(f"Scanning {DATA_DIR}...")
    files = get_files(DATA_DIR)
    total_files = len(files)
    print(f"Found {total_files} files to upload.")
    
    if total_files == 0:
        return

    completed = 0
    start_time = time.time()

    with concurrent.futures.ThreadPoolExecutor(max_workers=CONCURRENCY) as executor:
        # Submit all tasks
        future_to_file = {executor.submit(upload_file, f): f for f in files}
        
        for future in concurrent.futures.as_completed(future_to_file):
            completed += 1
            success, msg = future.result()
            
            # Calculate progress
            percent = (completed / total_files) * 100
            elapsed = time.time() - start_time
            rate = completed / elapsed if elapsed > 0 else 0
            
            # Print progress bar
            bar_length = 30
            filled_length = int(bar_length * completed // total_files)
            bar = '█' * filled_length + '-' * (bar_length - filled_length)
            
            sys.stdout.write(f'\r[{bar}] {percent:.1f}% ({completed}/{total_files}) - {msg if not success else "OK"}')
            sys.stdout.flush()
            
            if not success:
                print(f"\nError: {msg}")

    print(f"\n\nUpload complete! Total time: {time.time() - start_time:.2f}s")

if __name__ == "__main__":
    main()

File: backend/scripts/upload_parallel.js
const fs = require('fs');
const path = require('path');
const { exec } = require('child_process');

const DATA_DIR = path.join(__dirname, '../../data/chunks');
const BUCKET_NAME = 'austin-codes';
const CONCURRENCY = 10; // Number of parallel uploads

async function getFiles(dir) {
    const dirents = await fs.promises.readdir(dir, { withFileTypes: true });
    const files = await Promise.all(dirents.map((dirent) => {
        const res = path.resolve(dir, dirent.name);
        return dirent.isDirectory() ? getFiles(res) : res;
    }));
    return Array.prototype.concat(...files);
}

async function uploadFile(filePath) {
    const relativePath = path.relative(path.join(__dirname, '../../data'), filePath).replace(/\\/g, '/');
    const command = `raindrop object put "${filePath}" "${relativePath}" --bucket "${BUCKET_NAME}"`;

    return new Promise((resolve, reject) => {
        exec(command, (error, stdout, stderr) => {
            if (error) {
                console.error(`Failed to upload ${relativePath}: ${stderr}`);
                reject(error);
            } else {
                console.log(`Uploaded: ${relativePath}`);
                resolve();
            }
        });
    });
}

async function main() {
    console.log(`Scanning ${DATA_DIR}...`);
    const allFiles = await getFiles(DATA_DIR);
    console.log(`Found ${allFiles.length} files.`);

    const queue = [...allFiles];
    const active = new Set();

    while (queue.length > 0 || active.size > 0) {
        while (queue.length > 0 && active.size < CONCURRENCY) {
            const file = queue.shift();
            const promise = uploadFile(file).then(() => {
                active.delete(promise);
            }).catch(() => {
                active.delete(promise);
            });
            active.add(promise);
        }
        if (active.size > 0) {
            await Promise.race(active);
        }
    }
    console.log("All uploads complete!");
}

main();
